# ==============================================================================
# REGRESSOR COMMIT DIFF
# ==============================================================================
# File: third_party/rust/wgpu-core/src/command/encoder.rs
# Commit: 3f167f8b3137
# Full Hash: 3f167f8b313749f50f135e189e5854bef7ac0c73
# Author: Erich Gubler <erichdongubler@gmail.com>
# Date: 2025-10-03 18:08 +0000
# Regressor Bug: 1991551
# File Overlap Count: 23
# Overlapping Files: third_party/rust/wgpu-hal/Cargo.toml, supply-chain/audits.toml, third_party/rust/naga/src/back/spv/writer.rs, third_party/rust/naga/src/back/msl/writer.rs, third_party/rust/wgpu-types/src/lib.rs
# Description:
#   Bug 1991551 - build(webgpu): update WGPU to 8c4aebc0c1b75e3f7412dfc180745a0717fc30ec r=webgpu-reviewers,supply-chain-reviewers,nical
# ==============================================================================

diff -r 823439f307a0 -r 3f167f8b3137 third_party/rust/wgpu-core/src/command/encoder.rs
--- a/third_party/rust/wgpu-core/src/command/encoder.rs	Fri Oct 03 18:01:30 2025 +0000
+++ b/third_party/rust/wgpu-core/src/command/encoder.rs	Fri Oct 03 18:08:29 2025 +0000
@@ -1,22 +1,45 @@
 use alloc::{sync::Arc, vec::Vec};
 
 use crate::{
-    command::memory_init::CommandBufferTextureMemoryActions, device::Device,
-    init_tracker::BufferInitTrackerAction, ray_tracing::AsAction, snatch::SnatchGuard,
+    command::memory_init::CommandBufferTextureMemoryActions,
+    device::{queue::TempResource, Device},
+    init_tracker::BufferInitTrackerAction,
+    ray_tracing::AsAction,
+    snatch::SnatchGuard,
     track::Tracker,
 };
 
-/// State applicable when encoding commands onto a compute pass, or onto a
-/// render pass, or directly with a command encoder.
-pub(crate) struct EncodingState<'snatch_guard, 'cmd_enc, 'raw_encoder> {
+/// State applicable when encoding commands onto a compute pass, render pass, or
+/// directly to a command encoder.
+///
+/// Most encoding routines just want to receive an open encoder, write
+/// command(s) to it, and leave it open for whatever is next. In this case the
+/// `E` type parameter has the default value of `dyn hal::DynCommandEncoder`. To
+/// avoid confusion about encoder state, we set the convention that _the encoder
+/// in an `EncodingState` holding a bare HAL reference must always be open_.
+///
+/// Compute and render passes are more complicated. Because they record a
+/// command buffer for a housekeeping pre-pass which is inserted before the pass
+/// itself, the first thing they will do is close and reopen the encoder if it
+/// is already open. Unnecessary empty HAL passes can be avoided by passing them
+/// the encoder in whatever state it happens to be. In this case, `E` is
+/// `InnerCommandEncoder`, which tracks the state of the encoder. The callee
+/// (the render or compute pass) will open and close the encoder as necessary.
+///
+/// This structure is not supported by cbindgen because it contains a trait
+/// object reference.
+///
+/// cbindgen:ignore
+pub(crate) struct EncodingState<'snatch_guard, 'cmd_enc, E: ?Sized = dyn hal::DynCommandEncoder> {
     pub(crate) device: &'cmd_enc Arc<Device>,
 
-    pub(crate) raw_encoder: &'raw_encoder mut dyn hal::DynCommandEncoder,
+    pub(crate) raw_encoder: &'cmd_enc mut E,
 
     pub(crate) tracker: &'cmd_enc mut Tracker,
     pub(crate) buffer_memory_init_actions: &'cmd_enc mut Vec<BufferInitTrackerAction>,
     pub(crate) texture_memory_actions: &'cmd_enc mut CommandBufferTextureMemoryActions,
     pub(crate) as_actions: &'cmd_enc mut Vec<AsAction>,
+    pub(crate) temp_resources: &'cmd_enc mut Vec<TempResource>,
     pub(crate) indirect_draw_validation_resources:
         &'cmd_enc mut crate::indirect_validation::DrawResources,
 